### PR TITLE
fix: close post-merge reliability review gaps

### PR DESCRIPTION
**Pull request by Th0rgal**
## Summary

Closes the valid reliability findings left on the already-merged Hermes/remote-execution PR series (#641, #643–#657):

- preserve retryable Hermes messages after in-stream failures and stop stale 401 responses from clearing a newer dashboard login
- distinguish upstream Hermes authentication failures from dashboard authentication failures
- preserve Codex turn-outcome metadata when backend data is attached
- prioritize terminal remote-build wakes and close detached waiting leases safely
- recover the newest remote validation by submission sequence, not misleading completion timestamps
- return the documented synchronous response shape when an immutable remote validation receipt is reused
- fix alerts pagination exhaustion and compressed-session cron delivery
- tighten PR writer inference around policy adjectives such as “prohibited dependency”

## Verification

- `cargo fmt --all --check`
- `cargo check --all-targets`
- `cargo test --locked --workspace` (1,475 passed; 1 ignored)
- `cargo clippy --locked --workspace -- -D clippy::all`
- `scripts/harness_contract_tests.sh`
- `scripts/check-capability-matrix.sh`
- `cd dashboard && bun run test:unit` (245 passed)
- `cd dashboard && bun run lint` (no errors; pre-existing warnings only)
- `cd dashboard && bun run build`

---

> [!NOTE]
> **Medium Risk**
> Touches mission control queue ordering, remote-build lease lifecycle, JWT clearing, and Hermes SessionDB schema/delivery—important paths but changes are narrowly scoped with tests.
> 
> **Overview**
> Closes reliability gaps across the dashboard, control plane, Hermes delivery, and remote validation paths.
> 
> **Dashboard:** `apiFetch` only clears JWT and signals auth when the 401 matches the token that was sent, so a slow stale request cannot log out a newer session. Hermes chat streams that fail mid-turn now surface like other send errors (draft + Retry). **Alerts** pagination tracks a first-page anchor so a refresh can reopen “Load older” and merge/dedupe/sort correctly. The Hermes chat proxy maps upstream **401** to **502** so dashboard login is not blamed for bad `API_SERVER_KEY`.
> 
> **Backend:** `AgentResult::with_data` merges object fields so turn-outcome metadata is not wiped when extra diagnostics are attached. Remote-build terminal wakes are **enqueued at the front**; recovery picks continuations by **submission sequence**; parked missions can finish detached `waiting_remote_job` runs; PR-writer inference respects policy phrasing like “prohibited by policy” vs repair intents. Reused immutable remote validation receipts with `wait=true` return the documented wait response shape.
> 
> **Hermes patches:** API-server cron delivery resolves **resume/continuation** session IDs; durable delivery uses **delivery_id** receipts, bounded SessionDB write retries, **spool + gateway replay** on lock failure, and idempotent replay through resumed sessions.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7963ebf1a569096ebb2cfca9c89e9ef46bf129af. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by Th0rgal** [comment]
@codex review

**Comment by Th0rgal** [comment]
@codex review

File: dashboard/src/lib/api/core.ts [hunk 1]
```diff
@@ -70,13 +75,16 @@ export function isHttpStatusError(
...
   const headers: Record<string, string> = {
     ...(init?.headers ? (init.headers as Record<string, string>) : {}),
     ...authHeader(),
   };
 
   const res = await fetch(apiUrl(path), { ...init, headers });
-  if (res.status === 401) {
+  // A delayed 401 from a request carrying an old token must not erase a token
+  // issued by a newer successful login.
+  if (res.status === 401 && getStoredJwt()?.token === requestJwt) {
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/lib/api/core.ts]: Fixed in 1040af80. The current stored token is now normalized to `null` before comparison, so an unauthenticated/expired-token 401 still dispatches the auth-required event, while a delayed response carrying an older token cannot clear a newer login. Added `signals for a 401 when no login token was sent`; the targeted unit suite, lint, and production build pass.

**Comment by Th0rgal** [comment]
@codex review

File: dashboard/src/lib/api/core.ts [hunk 2]
```diff
@@ -70,13 +75,19 @@ export function isHttpStatusError(
 // ---------------------------------------------------------------------------
 
 export async function apiFetch(path: string, init?: RequestInit): Promise<Response> {
+  const requestJwt = getStoredJwt()?.token ?? null;
   const headers: Record<string, string> = {
     ...(init?.headers ? (init.headers as Record<string, string>) : {}),
     ...authHeader(),
   };
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/lib/api/core.ts]: Fixed in 600a7538. `authHeader()` now validates/clears expiration first, and the snapshot is parsed from the actual Bearer header that will be sent. Added an expired-token 401 regression test alongside the no-token and delayed-old-token cases; all 15 core API tests and the production dashboard build pass.

File: src/api/control/mod.rs [hunk 3]
```diff
@@ -5961,10 +5974,12 @@ fn inferred_pr_writer(explicit: Option<bool>, intent: Option<&str>, prompt: Opti
...
                 index >= 2 && words[index - 2] == "stop" && words[index - 1] == "before";
-            let after_end = (index + 4).min(words.len());
-            let negated_after = words[(index + 1).min(words.len())..after_end]
-                .iter()
-                .any(|next| NEGATORS_AFTER.contains(next));
+            let negated_after = words.get(index + 1).is_some_and(|word| {
+                matches!(*word, "is" | "are" | "was" | "were")
+                    && words
+                        .get(index + 2)
+                        .is_some_and(|next| NEGATORS_AFTER.contains(next))
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/control/mod.rs]: Fixed in 600a7538. Immediate prohibitions are preserved when terminal or followed by a policy qualifier (`Push forbidden`, `Merge prohibited by policy`), while object-adjective phrases such as `Fix the prohibited dependency usage` and `Implement the banned-import repair` remain writers. Added all four cases to the inference regression test; targeted Rust test and strict Clippy pass.

**Comment by Th0rgal** [comment]
@codex review

File: dashboard/src/components/hermes/alerts-feed.tsx [hunk 4]
```diff
@@ -61,12 +65,13 @@ export function AlertsFeed() {
     ];
   }, [firstPage, olderPages]);
 
-  const cursor = olderCursor ?? data?.next_cursor ?? null;
+  const cursor =
+    olderCursor === undefined ? (data?.next_cursor ?? null) : olderCursor;
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/components/hermes/alerts-feed.tsx]: Fixed in 165e3c52. Pagination now records the first-page boundary that produced its deeper/exhausted cursor. When SWR moves that boundary, the new gap cursor becomes reachable again; deduped alerts are sorted chronologically so a late gap fetch does not reorder the old tail. Added a refresh-after-exhaustion regression test; 248 dashboard tests, lint, and production build pass.

File: patches/hermes/api-server-session-cron-delivery.patch [hunk 5]
```diff
@@ -24,20 +24,21 @@ index c595707..9e25672 100644
 +                from hermes_state import SessionDB
 +
 +                db = SessionDB()
-+                session = db.get_session(str(chat_id))
++                resolved_chat_id = db.resolve_resume_session_id(str(chat_id))
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply patches/hermes/api-server-session-cron-delivery.patch:27]: Fixed in 165e3c52. Failed SessionDB appends now spool `resolved_chat_id`. I also regenerated the scheduler/test patch hunks with semantic context so `--unidiff-zero` cannot silently place them in unrelated code. Both managed patches apply sequentially to the exact Hermes baseline, `git diff --check` passes, and the origin-delivery plus spool suites pass (3 + 3 tests), including a parent-to-child spool assertion.

**Comment by Th0rgal** [comment]
@codex review

File: patches/hermes/sessiondb-durable-delivery.patch [hunk 6]
```diff
@@ -24,7 +46,7 @@ diff --git a/cron/scheduler.py b/cron/scheduler.py
 +
 +                    spool_path = spool_session_delivery(
 +                        delivery_id=delivery_id,
-+                        session_id=str(chat_id),
++                        session_id=resolved_chat_id,
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply patches/hermes/sessiondb-durable-delivery.patch:50]: Fixed in 0874d4e1. The scheduler initializes `resolved_chat_id` to the origin alias before entering the fallible DB path, then overwrites it with the live continuation when resolution succeeds. Spool replay now resolves any stored parent alias again after the DB recovers, so a pre-resolution lock failure remains durable and still lands on the continuation. The managed patches apply sequentially with contextual scheduler hunks; origin tests pass 3/3 and spool tests pass 4/4, including parent-to-child replay.

**Comment by Th0rgal** [comment]
@codex review

File: src/api/control/mod.rs [hunk 7]
```diff
@@ -5961,11 +5974,26 @@ fn inferred_pr_writer(explicit: Option<bool>, intent: Option<&str>, prompt: Opti
...
-            let after_end = (index + 4).min(words.len());
-            let negated_after = words[(index + 1).min(words.len())..after_end]
-                .iter()
-                .any(|next| NEGATORS_AFTER.contains(next));
-            !(negated_immediately || coordinated_negation || stopped_before || negated_after)
+            let copula_prohibition = words.get(index + 1).is_some_and(|word| {
+                matches!(*word, "is" | "are" | "was" | "were")
+                    && words
+                        .get(index + 2)
+                        .is_some_and(|next| NEGATORS_AFTER.contains(next))
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/control/mod.rs]: Fixed in 7963ebf1. Writer inference now performs a bounded post-verb scan: optional copula, up to two recognized policy adverbs, then the prohibition token. This covers `Merging is currently prohibited` and `Push is temporarily forbidden` while retaining writer classification for object phrases such as `Fix the prohibited dependency usage`. The expanded regression test and strict workspace Clippy pass.

**Comment by Th0rgal** [comment]
@codex review
